Source/BluetoothManager.swift

```diff
@@ -342,22 +342,29 @@ public class BluetoothManager {
     /// - Parameter peripheral: `Peripheral` which is monitored for connection.
     /// - Returns: Observable which emits next events when `peripheral` was connected.
     public func monitorConnection(for peripheral: Peripheral) -> Observable<Peripheral> {
-        return monitorPeripheral(on: delegateWrapper.rx_didConnectPeripheral, peripheral: peripheral)
+        return monitorPeripheral(
+            on: delegateWrapper.rx_didConnectPeripheral.map { ($0, nil) },
+            peripheral: peripheral
+        ).map { $0.0 }
     }
 
     /// Emits `Peripheral` instance when it's disconnected.
     /// - Parameter peripheral: `Peripheral` which is monitored for disconnection.
-    /// - Returns: Observable which emits next events when `peripheral` was disconnected.
-    public func monitorDisconnection(for peripheral: Peripheral) -> Observable<Peripheral> {
-        return monitorPeripheral(on: delegateWrapper.rx_didDisconnectPeripheral.map { $0.0 }, peripheral: peripheral)
+    /// - Returns: Observable which emits next events when `peripheral` was disconnected or error event when occured.
+    public func monitorDisconnection(for peripheral: Peripheral) -> Observable<(Peripheral, BluetoothError?)> {
+        return monitorPeripheral(
+            on: delegateWrapper.rx_didDisconnectPeripheral.map { (_, error) in
+                throw BluetoothError.peripheralDisconnectionFailed(peripheral, error)
```

Comment:
I might not read it correctly, but is there a reason for why you're ignoring the peripheral returned from `peripheralAction` observable here? It seems to me that the logic deciding which instance should be returned in the error should be kept in one place and `monitorPeripheral` is better place in my opinion.

Reply:
This code is not behaving correctly. We shouldn't throw error there.

Reply:
+1. In the issue I said that error should be emitted in the resulting observable next events, not as an error on stream.

Reply:
corrected throwing. It should be mapped as you mentioned.